File: applicatie/src/main/java/com/vdzon/administratie/afschrift/ImportFromAbnAmro.java
package com.vdzon.administratie.afschrift;


import com.vdzon.administratie.afschrift.regelParsers.DefaultRegelParser;
import com.vdzon.administratie.afschrift.regelParsers.RegelParser;
import com.vdzon.administratie.afschrift.regelParsers.SepaRegelParser;
import com.vdzon.administratie.afschrift.regelParsers.TRTPRegelParser;
import com.vdzon.administratie.model.Afschrift;
import com.vdzon.administratie.model.Gebruiker;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static com.vdzon.administratie.afschrift.ImportFromAbnAmroHelper.*;

public class ImportFromAbnAmro {


    public List<Afschrift> parseFile(Path out, Gebruiker gebruiker) {
        NextNummerHolder nextNummerHolder = new NextNummerHolder(findNextAfschriftNummer(gebruiker));
        try (Stream<String> stream = Files.lines(out)) {
            return stream.map(line -> parseLine(line, gebruiker, nextNummerHolder)).collect(Collectors.toList());
        } catch (IOException e) {
            e.printStackTrace();
            return new ArrayList<>();
        }
    }

    private Afschrift parseLine(String line, Gebruiker gebruiker, NextNummerHolder nextNummerHolder) {
        AfschriftData afschriftData = parseAfschriftData(line, str -> extractNaam(str), str -> extractOmschrijving(str));
        if (afschriftData == null) {
            return null;
        }
        Afschrift bestaandAfschrift = getBestaandAfschriftMetDezeUuid(gebruiker, afschriftData.uuid);
        if (bestaandAfschrift != null) {
            return null;
        }
        return buildAfschrift(nextNummerHolder, afschriftData);
    }

    private String extractNaam(String omschrijving) {
        RegelParser regel = getRegelParsers().stream().filter(parser -> parser.match(omschrijving)).findFirst().orElse(new DefaultRegelParser());
        return regel.extractNaam(omschrijving);
    }

    private String extractOmschrijving(String omschrijving) {
        RegelParser regel = getRegelParsers().stream().filter(parser -> parser.match(omschrijving)).findFirst().orElse(new DefaultRegelParser());
        return regel.extractOmschrijving(omschrijving);
    }

    private List<RegelParser> getRegelParsers() {
        List<RegelParser> parsers = new ArrayList<>();
        parsers.add(new SepaRegelParser());
        parsers.add(new TRTPRegelParser());
        return parsers;
    }
}

File: applicatie/src/main/java/com/vdzon/administratie/bankimport/ImportFromAbnAmro.kt
package com.vdzon.administratie.bankimport


import com.vdzon.administratie.bankimport.ImportFromAbnAmroHelper.buildAfschrift
import com.vdzon.administratie.bankimport.ImportFromAbnAmroHelper.findNextAfschriftNummer
import com.vdzon.administratie.bankimport.ImportFromAbnAmroHelper.getBestaandAfschriftMetDezeUuid
import com.vdzon.administratie.bankimport.ImportFromAbnAmroHelper.parseAfschriftData
import com.vdzon.administratie.bankimport.regelParsers.DefaultRegelParser
import com.vdzon.administratie.bankimport.regelParsers.RegelParser
import com.vdzon.administratie.bankimport.regelParsers.SepaRegelParser
import com.vdzon.administratie.bankimport.regelParsers.TRTPRegelParser
import com.vdzon.administratie.model.Afschrift
import com.vdzon.administratie.model.Gebruiker

import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.util.ArrayList
import java.util.stream.Collectors
import java.util.stream.Stream


class ImportFromAbnAmro {


    fun parseFile(out: Path, gebruiker: Gebruiker): List<Afschrift> {
        val nextNummerHolder = NextNummerHolder(findNextAfschriftNummer(gebruiker))
        try {
            return Files.lines(out).map({ line:String -> parseLine(line, gebruiker, nextNummerHolder) }).collect(Collectors.toList<Afschrift>())
        } catch (e: IOException) {
            e.printStackTrace()
            return ArrayList()
        }
    }

    private fun parseLine(line: String, gebruiker: Gebruiker, nextNummerHolder: NextNummerHolder): Afschrift? {
        val afschriftData = parseAfschriftData(line, { str:String -> extractNaam(str) } , { str -> extractOmschrijving(str) }) ?: return null

        val bestaandAfschrift = getBestaandAfschriftMetDezeUuid(gebruiker, afschriftData.uuid)
        if (bestaandAfschrift != null) {
            return null
        }
        return buildAfschrift(nextNummerHolder, afschriftData)
    }

    private fun extractNaam(omschrijving: String): String {
        val regel = regelParsers.filter{ parser -> parser.match(omschrijving) }.firstOrNull()?:DefaultRegelParser()
        return regel.extractNaam(omschrijving)
    }

    private fun extractOmschrijving(omschrijving: String): String {
        val regel = regelParsers.filter{ parser -> parser.match(omschrijving) }.firstOrNull()?:DefaultRegelParser()
        return regel.extractOmschrijving(omschrijving)
    }

    private val regelParsers: List<RegelParser>
        get() {
            val parsers = ArrayList<RegelParser>()
            parsers.add(SepaRegelParser())
            parsers.add(TRTPRegelParser())
            return parsers
        }
}